csi/volumeutils.py

```diff
@@ -27,962 +27,1361 @@
 VOLFILES_DIR = "/kadalu/volfiles"
 TEMPLATES_DIR = "/kadalu/templates"
-VOLINFO_DIR = "/var/lib/gluster"
+POOLINFO_DIR = "/var/lib/gluster"
+SECRET_PRIVATE_KEY = "/etc/secret-volume/ssh-privatekey"
 
 # This variable contains in-memory map of all glusterfs processes (hash of volfile and pid)
 # Used while sending SIGHUP during any modifcation to storage config
-VOL_DATA = {}
+POOL_DATA = {}
 
 statfile_lock = threading.Lock()    # noqa # pylint: disable=invalid-name
 mount_lock = threading.Lock()    # noqa # pylint: disable=invalid-name
+SINGLE_NODE_WRITER = getattr(csi_pb2.VolumeCapability.AccessMode,
+                             "SINGLE_NODE_WRITER")
 
+MULTI_NODE_MULTI_WRITER = getattr(csi_pb2.VolumeCapability.AccessMode,
+                                  "MULTI_NODE_MULTI_WRITER")
+LATEST_VOLUME_CONTEXT_VERSION = "2"
 
-class Volume():
-    """Hosting Volume object"""
-    def __init__(self, volname, voltype, hostvol, **kwargs):
-        self.voltype = voltype
-        self.volname = volname
-        self.volhash = kwargs.get("volhash", None)
-        self.volpath = kwargs.get("volpath", None)
-        self.hostvol = hostvol
-        self.size = kwargs.get("size", None)
-        self.extra = {}
-        self.extra['ghost'] = kwargs.get("ghost", None)
-        self.extra['hostvoltype'] = kwargs.get("hostvoltype", None)
-        self.extra['gvolname'] = kwargs.get("gvolname", None)
-        self.extra['kformat'] = kwargs.get("kformat", 'native')
-        self.setpath()
-
-    def setpath(self):
-        """Set Volume path based on hash and volume name"""
-        if self.volpath is None:
-            self.volpath = get_volume_path(
-                self.voltype,
-                self.volhash,
-                self.volname
-            )
 
-    def get(self):
-        """Get Volume name"""
-        return self.volname
+def pv_info_file_path(mountpath, pvtype, pvname):
+    """
+    PV info file path. Format: <mnt>/info/<path>
+    """
+    return f"{mountpath}/info/{pv_path(pvtype, pvname)}"
 
 
-def filter_node_affinity(volume, filters):
+def options_based_pool_mount_suffix(opts, mnt_opts):
     """
-    Filter volume based on node affinity provided
+    If additional Options provided via Storage Class then
+    generate a suffix that can be used while mounting the Pool.
+    This allows to have multiple mounts of the same Pool with
+    different Volfile Options or Mount Options.
     """
-    node_name = filters.get("node_affinity", None)
-    if node_name is not None:
-        # Node affinity is only applicable for Replica1 Volumes
-        if volume["type"] != "Replica1":
-            return None
+    all_opts = ""
+    if opts != "":
+        all_opts += opts
+    if mnt_opts != "":
+        all_opts += mnt_opts
 
-        # Volume is not from the requested node
-        if node_name != volume["bricks"][0]["kube_hostname"]:
-            return None
+    if all_opts != "":
+        return "_" + xxhash.xxh64_hexdigest(all_opts)
```

Comment:
This can be fine-tuned later. For now, if two Storage classes provide options like `log-level=DEBUG,opt1=val1` and `opt1=val1,log-level=DEBUG` will create two mounts in respective CSI node plugin pod.